From 80da2f511f6384c9f1ab120e172e8d91ad0b7603 Mon Sep 17 00:00:00 2001
From: nxf32288 <ganesh.deva_1@nxp.com>
Date: Tue, 12 Apr 2022 15:38:15 +0530
Subject: [PATCH 4/6] [artf955603]: Added support for Keymint 2.0

1) Default implementation done for newly added ROT related API's
2) RKP getHardwareInfo API aligned with KM 2.0 version.

Change-Id: Iab957f8fe2a51bdebf7147eebd1d5f44e2278b33
Signed-off-by: Ganesh Devaraj <ganesh.deva_1@nxp.com>
---
 keymint/JavacardKeyMintDevice.cpp             | 13 +++++++
 keymint/JavacardKeyMintDevice.h               |  7 ++++
 ...cardRemotelyProvisionedComponentDevice.cpp | 34 +++++++++++++++++++
 ...are.security.keymint-service.strongbox.xml |  2 ++
 .../android.hardware.strongbox_keystore.xml   |  2 +-
 5 files changed, 57 insertions(+), 1 deletion(-)

diff --git a/keymint/JavacardKeyMintDevice.cpp b/keymint/JavacardKeyMintDevice.cpp
index a645d16..03559ce 100644
--- a/keymint/JavacardKeyMintDevice.cpp
+++ b/keymint/JavacardKeyMintDevice.cpp
@@ -388,6 +388,19 @@ ScopedAStatus JavacardKeyMintDevice::getKeyCharacteristics(
     return ScopedAStatus::ok();
 }
 
+ScopedAStatus JavacardKeyMintDevice::getRootOfTrustChallenge(__attribute__((unused)) std::array<uint8_t, 16>* _aidl_return) {
+    return km_utils::kmError2ScopedAStatus(KM_ERROR_UNIMPLEMENTED);
+}
+
+ScopedAStatus JavacardKeyMintDevice::getRootOfTrust(__attribute__((unused)) const std::array<uint8_t, 16>& in_challenge,
+                                  __attribute__((unused)) std::vector<uint8_t>* _aidl_return) {
+    return km_utils::kmError2ScopedAStatus(KM_ERROR_UNIMPLEMENTED);
+}
+
+ScopedAStatus JavacardKeyMintDevice::sendRootOfTrust(__attribute__((unused)) const std::vector<uint8_t>& in_rootOfTrust) {
+    return km_utils::kmError2ScopedAStatus(KM_ERROR_UNIMPLEMENTED);
+}
+
 keymaster_error_t
 JavacardKeyMintDevice::parseWrappedKey(const vector<uint8_t>& wrappedKeyData,
                                        std::vector<uint8_t>& iv, std::vector<uint8_t>& transitKey,
diff --git a/keymint/JavacardKeyMintDevice.h b/keymint/JavacardKeyMintDevice.h
index 56cf662..1afaf6e 100644
--- a/keymint/JavacardKeyMintDevice.h
+++ b/keymint/JavacardKeyMintDevice.h
@@ -105,6 +105,13 @@ class JavacardKeyMintDevice : public BnKeyMintDevice {
     ScopedAStatus convertStorageKeyToEphemeral(const std::vector<uint8_t>& storageKeyBlob,
                                                std::vector<uint8_t>* ephemeralKeyBlob) override;
 
+    ScopedAStatus getRootOfTrustChallenge(std::array<uint8_t, 16>* _aidl_return) override;
+
+    ScopedAStatus getRootOfTrust(const std::array<uint8_t, 16>& in_challenge,
+                                  std::vector<uint8_t>* _aidl_return) override;
+
+    ScopedAStatus sendRootOfTrust(const std::vector<uint8_t>& in_rootOfTrust) override;
+
   private:
     keymaster_error_t parseWrappedKey(const vector<uint8_t>& wrappedKeyData,
                                       std::vector<uint8_t>& iv, std::vector<uint8_t>& transitKey,
diff --git a/keymint/JavacardRemotelyProvisionedComponentDevice.cpp b/keymint/JavacardRemotelyProvisionedComponentDevice.cpp
index 9055de9..b3b9a08 100644
--- a/keymint/JavacardRemotelyProvisionedComponentDevice.cpp
+++ b/keymint/JavacardRemotelyProvisionedComponentDevice.cpp
@@ -13,6 +13,25 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+/******************************************************************************
+ **
+ ** The original Work has been changed by NXP.
+ **
+ ** Licensed under the Apache License, Version 2.0 (the "License");
+ ** you may not use this file except in compliance with the License.
+ ** You may obtain a copy of the License at
+ **
+ ** http://www.apache.org/licenses/LICENSE-2.0
+ **
+ ** Unless required by applicable law or agreed to in writing, software
+ ** distributed under the License is distributed on an "AS IS" BASIS,
+ ** WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ ** See the License for the specific language governing permissions and
+ ** limitations under the License.
+ **
+ ** Copyright 2022 NXP
+ **
+ *********************************************************************************/
 
 #define LOG_TAG "javacard.keymint.device.rkp.strongbox-impl"
 #include <JavacardRemotelyProvisionedComponentDevice.h>
@@ -22,6 +41,10 @@
 #include <keymaster/cppcose/cppcose.h>
 #include <keymaster/remote_provisioning_utils.h>
 
+#ifdef NXP_EXTNS
+#define KM_RKP_VERSION_1 0x01
+#endif
+
 namespace aidl::android::hardware::security::keymint {
 using namespace cppcose;
 using namespace keymaster;
@@ -93,6 +116,17 @@ JavacardRemotelyProvisionedComponentDevice::getHardwareInfo(RpcHardwareInfo* inf
         return defaultHwInfo(info);
     }
     info->versionNumber = static_cast<int32_t>(versionNumber);
+#ifdef NXP_EXTNS
+    if (info->versionNumber > KM_RKP_VERSION_1) {
+        std::string uniqueId;
+        if (!cbor_.getBinaryArray(item, 4, uniqueId)) {
+            LOG(ERROR) << "Error in uniqueId response of getHardwareInfo.";
+            LOG(INFO) << "Returning defaultHwInfo in getHardwareInfo.";
+            return defaultHwInfo(info);
+        }
+        info->uniqueId = static_cast<::std::optional<::std::string>>(uniqueId);
+    }
+#endif
     info->supportedEekCurve = static_cast<int32_t>(supportedEekCurve);
     return ScopedAStatus::ok();
 }
diff --git a/keymint/android.hardware.security.keymint-service.strongbox.xml b/keymint/android.hardware.security.keymint-service.strongbox.xml
index c71f4a0..39524f3 100644
--- a/keymint/android.hardware.security.keymint-service.strongbox.xml
+++ b/keymint/android.hardware.security.keymint-service.strongbox.xml
@@ -1,10 +1,12 @@
 <manifest version="1.0" type="device">
     <hal format="aidl">
         <name>android.hardware.security.keymint</name>
+        <version>2</version>
         <fqname>IKeyMintDevice/strongbox</fqname>
     </hal>
     <hal format="aidl">
         <name>android.hardware.security.keymint</name>
+        <version>2</version>
         <fqname>IRemotelyProvisionedComponent/strongbox</fqname>
     </hal>
 </manifest>
diff --git a/keymint/android.hardware.strongbox_keystore.xml b/keymint/android.hardware.strongbox_keystore.xml
index d92d605..c6ca188 100644
--- a/keymint/android.hardware.strongbox_keystore.xml
+++ b/keymint/android.hardware.strongbox_keystore.xml
@@ -12,6 +12,6 @@
 -->
 <!-- Feature for devices with Keymaster in StrongBox. -->
 <permissions>
-  <feature name="android.hardware.strongbox_keystore" version="100"/>
+  <feature name="android.hardware.strongbox_keystore" version="200"/>
   <feature name="android.hardware.keystore.app_attest_key" />
 </permissions>
-- 
2.35.1

